docs/specs/environment-specific-sms-gateway.md

```diff
@@ -0,0 +1,141 @@
+# Environment specific SMS Gateway
+
+This document describes configuration of sms gateway.
+
+- [Using environment specific sms gateway](#using-environment-specific-sms-gateway)
+  - [Configuration](#configuration)
+- [Using Custom SMS Gateway](#using-custom-sms-gateway)
+  - [Configuration](#custom-sms-gateway-configuration)
+  - [Webhook](#webhook)
+  - [Deno Hook](#deno-hook)
+
+## Using environment specific sms gateway
+
+### Configuration
+
+The following configs must be specified in `.env`:
+
+```
+SMS_GATEWAY_TWILIO_ACCOUNT_SID=
+SMS_GATEWAY_TWILIO_AUTH_TOKEN=
+SMS_GATEWAY_TWILIO_MESSAGING_SERVICE_SID=
+
+SMS_GATEWAY_NEXMO_API_KEY=
+SMS_GATEWAY_NEXMO_API_SECRET=
+
+SMS_GATEWAY_CUSTOM_URL=
+SMS_GATEWAY_CUSTOM_TIMEOUT=
+
+SMS_GATEWAY_DEFAULT_CONFIG_FROM=env|messaging.sms_provider
+SMS_GATEWAY_DEFAULT_PROVIDER=twilio|nexmo|custom
+```
+
+- `SMS_GATEWAY_DEFAULT_CONFIG_FROM`
+  - The default sms gateway config source. Can be `env` or `messaging.sms_provider`.
+  - Default is `messaging.sms_provider` if missing.
+- `SMS_GATEWAY_DEFAULT_PROVIDER`
+  - The default sms provider.
+  - Default is `twilio` if missing.
+
+The following configs must be specified in `authgear.yaml`:
+
+```diff
+ messaging:
+   sms_provider: twillio | nexmo | custom
++  sms_gateway:
++    config_from: env | messaging.sms_provider
++    provider: twillio | nexmo | custom
+```
+
+- `messaging.sms_gateway`
+  - The new sms gateway config, capable to specify reading config from env or project config.
+  - If missing, the sms provider config will take `messaging.sms_provider` from `authgear.yaml` and `sms.*` from `authgear.secrets.yaml`.
+- `messaging.sms_gateway.config_from`
+  - Can be from `env` / `messaging.sms_provider`.
+    - `env`: Use config from `SMS_GATEWAY_*` environment variables.
+    - `messaging.sms_provider`: Use config from `messaging.sms_provider` from authgear.yaml and `sms.*` from `authgear.secrets.yaml`.
+  - If missing, will use `SMS_GATEWAY_DEFAULT_CONFIG_FROM` environment variable.
```

Comment:
No, it cannot be missing. If `sms_gateway` is defined, then the two fields must be present.